Integration of SignIn and SignUp with node client. Create mock my-user-context for storybook.

RaghvindYadav opened the pull request:
<!-- This template is to serve as a comprehensive outline for documenting updates to the code. Some of these sections may not be relevant to your PR. In that case, document it inline with `N/A`  -->

## What does this PR do?
This PR have few implementations or integrations of the following things:
- Sign In
- Sign Up
- Mock My User Context
- Some improvement in storybook for signIn and signUp page

### Related Issues & PRs
These PR is integration of server for some fronted issues and those are closed.

#### Key Decisions
<!-- Explain the technical decisions and alternative approaches considered for your new implementations and refactored solutions -->
- **Implemeneted Approach**:
  We use node client to integrate our server.
- **Alternatives Considered**:
  <!-- What else did you consider? -->
- **Rationale**:
  <!-- Why did you decide this was the most appropriate solution? -->

## Breaking Changes
<!-- Do others need to do something to successfully incorporate your changes into their work? -->
- [x] This PR includes breaking changes
  Need to do rebase if someone created a branch before it merge. 


### Next Steps
<!-- What needs to happen after this PR is merged? -->
- [ ] Follow-up tasks are created
  - [ ] Reset password
  - [ ] Reset password in storybook
  - [ ] Add more components in storybook and also of `shadcn` which seems to be necessary.

### How to Test
To test storybook, you simply need to run `pnpm run storybook` in the project terminal.

## Final Checklist
- [ ] Linked related issues
- [x] Rebased with target branch
- [x] No lint errors
- [x] QA'd in multiple browsers
- [x] Added/updated tests (if applicable)
- [x] Mobile responsive (if applicable)
- [x] Updated documentation (if applicable)
- [x] Notified the Team of this updates readiness

---

## AI Usage
<!-- Document any AI tools you used to develop this -->
- [ ] AI tools were used in development
  - **Tool used**:
    Augment code: I used to create mock user context, but I was not so much correct.
  - **Prompts used**:
    Write a my-user-mock-context user for story book and follow the my-user-context (I passed local directory of it).
- [ ] No AI tools were used
  - **Reason**:
    There is no more context about the mock providers or context, so I did some search and find out the solution to use it. I used my previous skills (Flutter mockito) to do that.

## References
- To write mock context: https://react.dev/learn/passing-data-deeply-with-context
- To write correct validations: https://github.com/baragaun/first-spark-ionic-app/blob/main/src/helpers/validationHelpers.ts
- To write mock functions: https://storybook.js.org/docs/writing-stories/mocking-data-and-modules/mocking-network-requests

Commit:
signup and signin

Commit:
listener fix

Commit:
validation and identType

Commit:
fixes the context calls

Commit:
storybook update

SignIn and SignUp flows are working, but still need to do some more stuffs.

Commit:
quick fix

Commit:
Delete bundles.json

jonahtwalker commented on src/lib/components/auth-card.svelte:
👍🏼

jonahtwalker committed:
- fix isAuthenticated flag
- renamed EmailVerification to RegisterEmailForm
- cannot run lint

jonahtwalker commented on src/stories/mocks/mock-user-context.ts:
```diff
@@ -0,0 +1,457 @@
+  isInitialized = true;
+  myUserId = '';
+  // Add the missing _listeners property
+  private _listeners: Record<string, Array<any>> = {};
```
I am getting an error here for this `any`, and it is blocking the `lint` script.

jonahtwalker commented on src/stories/pages/SignUpPage.stories.svelte:
```diff
@@ -13,52 +14,92 @@
-      const identifierInput = canvas.getByLabelText(/Email or Username/i);
-      await userEvent.type(identifierInput, 'test@example.com');
+    // Step 1: Enter email
+    const emailInput = canvas.getByLabelText(/Email/i);
```
This one is outputting an error in the Storybook console

jonahtwalker reviewed (commented):
Thanks Raghav, this is looking a lot better. I've been trying to test out the functionality between running the app and storybook. I've found a few unhappy paths we'll want to address as well that I'll include as screenshots.

When I go to sign up, I immediately fail due to `"Error: nodemailer not available"` from the server. Is there a config step I missed somewhere, or maybe another branch I should be on? I end up partially "authenticated" because I do get a user, but fail the multistep flow. I can still verify with `666666`. When I create the credentials, I end up at Sign Up seemingly unauthenticated, but I am already logged in.

When I go to sign in, I fail the multistep flow but can validate with `666666`. Nothing seems to happen.

Will follow up with another review and more comments later tonight.

jonahtwalker commented:
These are some unhappy paths I came across, and some other QA finds. I hope these may help identify some edge cases and work out some of the quirks with the features!

- [x] I do think that we should refactor from onblur to a debounce. I did come across `debounce-throttle` when looking at Superforms examples, but I think there is a built in method in that library, too.

Can submit without a valid email in sign up
![Screenshot 2025-03-26 at 3 45 12 PM](https://github.com/user-attachments/assets/f94ce96a-173e-42d7-8ad5-2de2862e718c)

This seems like an error message that should be shown, but it actually only appears on hover with an error in the input
![Screenshot 2025-03-26 at 3 39 52 PM](https://github.com/user-attachments/assets/0eeea076-8f46-4366-b731-6462694be77e)

The sign in link is failing because the username does not exist. The message should also refer to the verification code, not a "magic link".
![Screenshot 2025-03-26 at 3 39 31 PM](https://github.com/user-attachments/assets/9f97d306-ec95-4b68-b0bf-8a23b9a357a0)

- Can sign in with an email that does not exist (while this seems okay, we should redirect to signup instead)

- The verification code seems to be verified every time in stories (success message comes in red text)
![Screenshot 2025-03-26 at 8 53 34 PM](https://github.com/user-attachments/assets/da6960be-240d-4d8a-a816-30075f0d481b)

- I do get a success message in the console when I verify with the test code, but I cannot proceed. (I can get this to fail more consistently than I can view the credentials-form.
![Screenshot 2025-03-26 at 8 56 22 PM](https://github.com/user-attachments/assets/2020c1b2-d713-4c4b-a968-69bc4de1f0e3)

@RaghvindYadav I hope my notes are beneficial, I think you're getting close to wrapping this chunk of work up! Is it possible that I missed some instruction or a note somewhere? I will follow up in the morning with Holger and see if maybe there's something in the backend that I need to change. I do think the nodemailer error is the crux of my issue, so hopefully it's an easy fix that I'm just overlooking.

The code is looking much much better!

RaghvindYadav commented:
> When I go to sign up, I immediately fail due to `"Error: nodemailer not available"` from the server. Is there a config step I missed somewhere, or maybe another branch I should be on? 

You need some more credentials from Holger it's related to server issue (env missing). Holger doesn't share with me too.

> I end up partially "authenticated" because I do get a user, but fail the multistep flow. I can still verify with `666666`. 

I need more context for this to understand.

> When I create the credentials, I end up at Sign Up seemingly unauthenticated, but I am already logged in.

Yes. it is because of Holger outliers. He suggest to do signup using email only than do whatever steps you need.  

> When I go to sign in, I fail the multistep flow but can validate with `666666`. Nothing seems to happen.

If you have created new account using first-spark-app then it should `666666` should work for every where except in storybook there `123456` confirmation code.

RaghvindYadav commented:
> The verification code seems to be verified every time in stories (success message comes in red text)

It's true some how page not jumping to home page because of story book limitation. I should not need to call home page manually, listener will call it. I am looking solution for this.

Commit:
bug fixes

improve the test cases for signIn and signUp

create a new components for ident input and used debounce

jonahtwalker committed:
- normalized use of myUserContext
- fixed isAuthenticated
- update myUser when user is found
- made signin/signup the same size card

jonahtwalker committed:
- refactored ident-input to accept current ident without error
- added loadUser to signIn methods
- fixed the credential form submission input
- refactored handleSignupSubmit to wait to submti

jonahtwalker committed:
- authenticated state needs to be simplified
- signing out does not update the UI state

jonahtwalker reviewed (commented):
Hey Raghav, thanks for the updates! I have made a few of my own, and I am primarily facing an issue with the signMeOut logic. For some reason it is not updating the UI auth state, where a refresh solves the issue. At this point I'm sure I've just looked past it so many times I'm not going to see it today, maybe you can find it? This authstate stuff should be cleaned up some more, and we should rely on the client's `isSignedIn` as the **truth**. `_authState` that I introduced may not be necessary.

In short, my changes:

- normalized use of myUserContext 
- update myUser when user is found
- made signin/signup the same size card
- refactored ident-input to accept the current ident without raising an error
- added loadUser to signInWithToken success case
- fixed the credential form submission input to actually submit the values
- refactored handleSignupSubmit to wait for submission
- random other little things that I wanted to see update based on auth

With the latest changes you can:
- complete the full sign up flow with a password and username update
- expect to be properly authenticated when you do sign up
- sign in with an email and test token (before the promise times out)
- expect to be properly authenticated when you sign in with token
- sign in with an email or username and password
- expect to be properly authenticated when you sign in with password
- expect your auth state to survive a hard refresh
- **expect signMeOut to not properly update the UI**
  - You do lose the auth token as you should
  - isSignedIn() returns true after signMeOut()
  - without a refresh the client will throw errors if called
  - a refresh properly de-authenticates you in the UI
  
  ---
  
https://github.com/user-attachments/assets/e3fe5ec6-30fc-4967-ab41-e83ba65ca830

---

@rupalirajput2597 If this branch is still not merged by the time you revisit the forms, you could base a branch off of this that actually updates the username, password or email. 👍🏼 Hope you're well!

RaghvindYadav commented:
Thanks, Jonah. I'll review your changes, and I'm sure I'll learn something from them. I hadn't planned to work on the navbar since I was dealing with the signOut issue. I was unsure if it was a backend problem, especially since you're encountering the same issue.

jonahtwalker commented:
> Thanks, Jonah. I'll review your changes, and I'm sure I'll learn something from them. I hadn't planned to work on the navbar since I was dealing with the signOut issue. I was unsure if it was a backend problem, especially since you're encountering the same issue.

Ah, thanks for the reminder. The NavBar needs a little tweaking to get the buttons right anyways, I'll revisit that part tomorrow. Thanks for checking out signout!

jonahtwalker committed:
- fix sign in authstate

jonahtwalker committed:
- minor refactoring
- signmeout does not work as expected

RaghvindYadav commented:
> Ah, thanks for the reminder. The NavBar needs a little tweaking to get the buttons right anyways, I'll revisit that part tomorrow. Thanks for checking out signout!

After looking a lot, I found sign out the issue and it is coming from node client. 
@shadowbrush created a function [clearMyUserFromClientInfo](https://github.com/baragaun/bg-node-client/blob/21c237defc6ba69dd5778d8e4edb36ed09aff659/src/helpers/clearMyUserFromClientInfo.ts) which helps us to modify the clientInfo via calling [persistClinetInfo](https://github.com/baragaun/bg-node-client/blob/21c237defc6ba69dd5778d8e4edb36ed09aff659/src/helpers/persistClientInfo.ts) method. I did not understand yet how it works. My eyes was rolling after think a lot on this.

I tried a simple way to `clientInfoStore.persist()` function to change the client info, I made it parallel to signInUser method and it worked.
```
const clearMyUserFromClientInfo = async (
  signedOutUserId?: string,
): Promise<ClientInfo> => {
  const clientInfo = await loadClientInfo();
  signedOutUserId = signedOutUserId || clientInfo.signedOutUserId;

  // Prepare the update object
  const updateInfo: Partial<ClientInfo> = {
    myUserId: null,
    authToken: null,
    signedOutUserId: signedOutUserId || clientInfo.signedOutUserId
  };

  console.log('clearMyUserFromClientInfo', updateInfo);

  // Use clientInfoStore.persist instead of directly calling persistClientInfo
  return clientInfoStore.persist(updateInfo);
};

export default clearMyUserFromClientInfo;
```

https://github.com/user-attachments/assets/9692947a-1f9f-4a81-bb9d-ca52a155b113


@shadowbrush I’m not sure why the node-client v18 caused issues with the signIn and signUp functions. I tried to fix it by referencing code from the git history and used this particular commit (21c237defc6ba69dd5778d8e4edb36ed09aff659).

Commit:
format and timeout

jonahtwalker commented on src/lib/components/ident-input.svelte:
```diff
@@ -1,14 +1,15 @@
   let debounceTimer: number | null = null;
-  // let identType = $state(UserIdentType.email);
   const DEBOUNCE_DELAY = 300; // ms
+  const myUserContext = getContext<MyUserContext>('myUserContext');
```
@RaghvindYadav Why can't you use `myUserContext` from the context directly?

Commit:
client config

Commit:
client config

Commit:
context

Commit:
* Upgraded `bg-node-client` to x19
* Changed MyUserContext.isSignedIn

Commit:
Fixed auth status

Commit:
Setting myUser

Commit:
myUser handling

Commit:
myUser handling

Commit:
Upgraded to 0.0.21

Commit:
listeners and bugs fixed

Except Forgot password, I catch it in next PR.

Commit:
earlier call for intialization

jonahtwalker committed:
run format
